components/nodemanager-service/pgdb/project_update.go

```diff
@@ -285,3 +193,185 @@ func (node *NodeProjectData) getValues(valuesType string) []string {
+		prefix := make([]byte, 1)
+		prefix[0] = uuidChars[i]
+		prefixes[i] = string(uuidChars[i])
+	}
+	return prefixes
+}
+
+func (db *DB) ListProjectUpdateTasks(ctx context.Context) ([]project_update_lib.SerializedProjectUpdateTask, error) {
+
+	var count int
+	row := db.Db.QueryRow(countNodesProjectData)
+	if err := row.Scan(&count); err != nil {
+		return nil, err
+	}
+
+	if count <= 50000 {
+		tasks := make([]project_update_lib.SerializedProjectUpdateTask, 1)
+		tasks[0] = project_update_lib.SerializedProjectUpdateTask{
+			Priority: 1 << 62,
+			Params: map[string]string{
+				"s": "0",
+				"e": "g",
+			},
+		}
+		return tasks, nil
+	}
+
+	prefixes := uuidPrefixRanges16()
+	logrus.Infof("Found %d nodes, splitting into %d chunks", count, len(prefixes))
+	tasks := make([]project_update_lib.SerializedProjectUpdateTask, len(prefixes))
+	for i := 0; i < len(prefixes); i++ {
+		start := prefixes[i]
+		end := ""
+		if i+1 < len(prefixes) {
+			end = prefixes[i+1]
+		} else {
+			end = "g"
+		}
+
+		tasks[i] = project_update_lib.SerializedProjectUpdateTask{
+			Priority: 1 << 62,
+			Params: map[string]string{
+				"s": start,
+				"e": end,
+			},
+		}
+	}
+	return tasks, nil
+
+}
+
+func (db *DB) RunProjectUpdateTask(ctx context.Context, projectUpdateID string,
+	params map[string]string, projectTaggingRules map[string]*iam_v2.ProjectRules) (
+	project_update_lib.SerializedProjectUpdateTaskID,
+	project_update_lib.SerializedProjectUpdateTaskStatus,
+	error) {
+
+	start := params["s"]
+	end := params["e"]
+
+	logrus.WithFields(logrus.Fields{
+		"start": start,
+		"end":   end,
+	}).Info("Running project update task for nodemanager")
+
+	if start == "" || end == "" {
+		return "", project_update_lib.SerializedProjectUpdateTaskStatus{}, errors.New("Invalid params")
+	}
+
+	iter := nodeDataIterator{
+		rangeStart: start,
+		rangeEnd:   end,
+		db:         db.DbMap,
+	}
+
+	for {
+		nodes, err := iter.Next()
+		if err != nil {
+			if err == io.EOF {
+				logrus.WithFields(logrus.Fields{
+					"start": start,
+					"end":   end,
+				}).Info("project update task complete for nodemanager")
+				return "", project_update_lib.SerializedProjectUpdateTaskStatus{
+					State:              project_update_lib.SerializedProjectUpdateTaskSuccess,
+					PercentageComplete: 1.0,
+				}, nil
+
+			}
+			return "", project_update_lib.SerializedProjectUpdateTaskStatus{}, err
+		}
+
+		if err := db.updateNodes(nodes, projectTaggingRules); err != nil {
+			return "", project_update_lib.SerializedProjectUpdateTaskStatus{}, err
+		}
+	}
+
+}
+
+func (db *DB) MonitorProjectUpdateTask(ctx context.Context,
+	projectUpdateID string, id project_update_lib.SerializedProjectUpdateTaskID) (
+	project_update_lib.SerializedProjectUpdateTaskStatus, error) {
+
+	return project_update_lib.NoStatus, errors.New("Unimplemented")
+
+}
+
+func (db *DB) CancelProjectUpdateTask(ctx context.Context, projectUpdateID string,
+	id project_update_lib.SerializedProjectUpdateTaskID) error {
```

Comment:
There is no way for this function to get called because the Run is synchronous. This function can only be called if Run has completed and it returned an id. If you cancel while Run is running, it has to wait for the Run to complete and then will cancel.

So the side effect here is that the portion of work we told the project update to do for nodemanager runs to completion before being canceled. We could improve this if we told the workflow it could just end if its running the nodemanager project update and it gets a cancel. If we thread context through to the right places in the nodemanager code, it would stop.
I didn't do it because it involved updating the workflow and my hope was it would complete in a reasonable amount of time